File: tools/cors-proxy.js
#!/usr/bin/env node
/* idg-cors-proxy — a tiny, zero-dependency local proxy that lets the
   browser-based generator reach an internal OpenAI-compatible endpoint
   (OpenShift AI / KServe-vLLM, etc.) WITHOUT launching Chrome with
   --disable-web-security.
 *
 * Why this is needed: the generator runs in the browser, and the browser
 * blocks cross-origin responses that don't carry CORS headers. OpenShift
 * Routes don't send those headers, so the call fails. This proxy sits on
 * localhost, forwards every request to your real endpoint, and adds the
 * missing CORS headers to the response — so the browser is satisfied.
 *
 * Usage:
 *   node tools/cors-proxy.js --upstream https://<model>-<project>.apps.<cluster> [options]
 *
 * Options:
 *   --upstream <url>   Required. Your real endpoint's ORIGIN (scheme://host[:port]),
 *                      i.e. the part of your working URL before "/v1".
 *                      A path prefix is allowed and will be prepended.
 *   --port <n>         Local port to listen on (default 8008).
 *   --host <addr>      Local bind address (default 127.0.0.1).
 *   --insecure         Skip TLS certificate verification (for clusters using an
 *                      internal/self-signed CA the machine doesn't trust).
 *   --origin <o>       Value for Access-Control-Allow-Origin (default "*").
 *
 * Then in the generator: Settings -> Connection
 *   Base URL  = http://localhost:8008/v1
 *   Model     = your model name
 *   API key   = your token (if the deployment requires one)
 *
 * No internet access is used; this only relays to the --upstream you provide.
 */
'use strict';

const http = require('http');
const https = require('https');
const { URL } = require('url');

function parseArgs(argv) {
  const o = { port: 8008, host: '127.0.0.1', insecure: false, origin: '*' };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--upstream') o.upstream = argv[++i];
    else if (a === '--port') o.port = parseInt(argv[++i], 10);
    else if (a === '--host') o.host = argv[++i];
    else if (a === '--origin') o.origin = argv[++i];
    else if (a === '--insecure') o.insecure = true;
    else if (a === '--help' || a === '-h') o.help = true;
  }
  return o;
}

// Headers we must not forward verbatim to the upstream.
const HOP = new Set(['host', 'origin', 'referer', 'connection', 'content-length', 'accept-encoding']);

/* Build an http.Server (not yet listening) that proxies to opts.upstream and
   adds CORS headers. Exported so the test harness can drive it directly. */
function createProxyServer(opts) {
  const upstream = new URL(opts.upstream);
  const upstreamLib = upstream.protocol === 'https:' ? https : http;
  const prefix = upstream.pathname.replace(/\/$/, ''); // '' when just an origin
  const allowOrigin = opts.origin || '*';
  const agent = upstream.protocol === 'https:'
    ? new https.Agent({ rejectUnauthorized: !opts.insecure, keepAlive: true })
    : new http.Agent({ keepAlive: true });

  function corsHeaders(req) {
    const reqHeaders = req.headers['access-control-request-headers'];
    return {
      'Access-Control-Allow-Origin': allowOrigin === '*' && req.headers.origin ? req.headers.origin : allowOrigin,
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': reqHeaders || 'Authorization, Content-Type',
      'Access-Control-Max-Age': '86400',
      'Vary': 'Origin',
    };
  }

  return http.createServer((req, res) => {
    const cors = corsHeaders(req);

    if (req.method === 'OPTIONS') {            // CORS preflight
      res.writeHead(204, cors);
      return res.end();
    }

    const targetPath = prefix + req.url;
    const headers = {};
    for (const [k, v] of Object.entries(req.headers)) {
      if (!HOP.has(k.toLowerCase())) headers[k] = v;
    }

    const upReq = upstreamLib.request({
      protocol: upstream.protocol,
      hostname: upstream.hostname,
      port: upstream.port || (upstream.protocol === 'https:' ? 443 : 80),
      method: req.method,
      path: targetPath,
      headers,
      agent,
    }, (upRes) => {
      // Relay upstream status + headers, then overlay CORS (so the browser
      // accepts it). Streaming bodies (SSE) are piped through unchanged.
      const outHeaders = Object.assign({}, upRes.headers, cors);
      res.writeHead(upRes.statusCode || 502, outHeaders);
      upRes.pipe(res);
    });

    upReq.on('error', (err) => {
      res.writeHead(502, Object.assign({ 'Content-Type': 'application/json' }, cors));
      res.end(JSON.stringify({ error: 'proxy_upstream_error', detail: String(err.message || err) }));
    });

    req.pipe(upReq);
  });
}

function main() {
  const opts = parseArgs(process.argv);
  if (opts.help || !opts.upstream) {
    console.log('Usage: node tools/cors-proxy.js --upstream https://<host> [--port 8008] [--host 127.0.0.1] [--insecure] [--origin "*"]');
    process.exit(opts.help ? 0 : 1);
  }
  try { new URL(opts.upstream); }
  catch (e) { console.error('Invalid --upstream URL: ' + opts.upstream); process.exit(1); }

  const server = createProxyServer(opts);
  server.listen(opts.port, opts.host, () => {
    const up = new URL(opts.upstream);
    console.log(`idg-cors-proxy → forwarding http://${opts.host}:${opts.port}  ⟶  ${up.origin}${up.pathname.replace(/\/$/, '')}`);
    console.log(`TLS verification: ${opts.insecure ? 'OFF (--insecure)' : 'on'}`);
    console.log('In the generator set Base URL to:  ' +
      `http://${opts.host === '0.0.0.0' ? 'localhost' : opts.host}:${opts.port}/v1`);
  });
}

if (require.main === module) main();

module.exports = { createProxyServer, parseArgs };

File: tools/cors_proxy.py
#!/usr/bin/env python3
"""idg-cors-proxy (Python stdlib, no dependencies) — lets the browser-based
generator reach an internal OpenAI-compatible endpoint (OpenShift AI /
KServe-vLLM, etc.) WITHOUT launching Chrome with --disable-web-security.

The generator runs in the browser, which blocks cross-origin responses that
lack CORS headers. OpenShift Routes don't send them, so the call fails. This
proxy runs on localhost, forwards every request to your real endpoint, and
adds the missing CORS headers — so the browser accepts the response.

Usage:
  python3 tools/cors_proxy.py --upstream https://<model>-<project>.apps.<cluster> [options]

Options:
  --upstream URL   Required. Your endpoint's origin (scheme://host[:port]) —
                   the part of your working URL before "/v1". A path prefix is
                   allowed and is prepended to forwarded requests.
  --port N         Local port (default 8008).
  --host ADDR      Local bind address (default 127.0.0.1).
  --insecure       Skip TLS certificate verification (internal/self-signed CA).
  --origin O       Access-Control-Allow-Origin value (default "*").

Then in the generator (Settings -> Connection):
  Base URL = http://localhost:8008/v1 ; Model = your model ; API key = token if needed

No internet access is used; it only relays to the --upstream you provide.
"""
import argparse
import http.client
import ssl
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse

HOP = {"host", "origin", "referer", "connection", "content-length", "accept-encoding"}


def build_handler(opts):
    up = urlparse(opts.upstream)
    prefix = up.path.rstrip("/")
    is_https = up.scheme == "https"
    port = up.port or (443 if is_https else 80)
    ctx = None
    if is_https:
        ctx = ssl.create_default_context()
        if opts.insecure:
            ctx.check_hostname = False
            ctx.verify_mode = ssl.CERT_NONE

    class Handler(BaseHTTPRequestHandler):
        protocol_version = "HTTP/1.1"

        def _allow_origin(self):
            o = self.headers.get("Origin")
            return o if (opts.origin == "*" and o) else opts.origin

        def _send_cors(self):
            self.send_header("Access-Control-Allow-Origin", self._allow_origin())
            self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
            self.send_header(
                "Access-Control-Allow-Headers",
                self.headers.get("Access-Control-Request-Headers", "Authorization, Content-Type"),
            )
            self.send_header("Access-Control-Max-Age", "86400")
            self.send_header("Vary", "Origin")

        def do_OPTIONS(self):
            self.send_response(204)
            self._send_cors()
            self.send_header("Content-Length", "0")
            self.end_headers()

        def _relay(self):
            body = None
            n = int(self.headers.get("Content-Length", 0) or 0)
            if n:
                body = self.rfile.read(n)
            fwd = {k: v for k, v in self.headers.items() if k.lower() not in HOP}
            conn = (http.client.HTTPSConnection(up.hostname, port, context=ctx)
                    if is_https else http.client.HTTPConnection(up.hostname, port))
            try:
                conn.request(self.command, prefix + self.path, body=body, headers=fwd)
                resp = conn.getresponse()
                self.send_response(resp.status)
                for k, v in resp.getheaders():
                    if k.lower() in ("transfer-encoding", "connection", "content-length",
                                     "access-control-allow-origin"):
                        continue
                    self.send_header(k, v)
                self._send_cors()
                self.send_header("Transfer-Encoding", "chunked")
                self.end_headers()
                # Stream upstream → client in chunks (keeps SSE responsive).
                while True:
                    chunk = resp.read(8192)
                    if not chunk:
                        break
                    self.wfile.write(b"%X\r\n%s\r\n" % (len(chunk), chunk))
                    self.wfile.flush()
                self.wfile.write(b"0\r\n\r\n")
            except Exception as e:  # noqa: BLE001
                msg = ('{"error":"proxy_upstream_error","detail":"%s"}' % str(e)).encode()
                self.send_response(502)
                self._send_cors()
                self.send_header("Content-Type", "application/json")
                self.send_header("Content-Length", str(len(msg)))
                self.end_headers()
                self.wfile.write(msg)
            finally:
                conn.close()

        do_GET = _relay
        do_POST = _relay

        def log_message(self, fmt, *args):  # quieter logging
            sys.stderr.write("  %s %s\n" % (self.command, self.path))

    return Handler


def main():
    ap = argparse.ArgumentParser(add_help=True)
    ap.add_argument("--upstream", required=True)
    ap.add_argument("--port", type=int, default=8008)
    ap.add_argument("--host", default="127.0.0.1")
    ap.add_argument("--origin", default="*")
    ap.add_argument("--insecure", action="store_true")
    opts = ap.parse_args()

    httpd = ThreadingHTTPServer((opts.host, opts.port), build_handler(opts))
    up = urlparse(opts.upstream)
    print("idg-cors-proxy → forwarding http://%s:%d  ⟶  %s://%s%s"
          % (opts.host, opts.port, up.scheme, up.netloc, up.path.rstrip("/")))
    print("TLS verification: %s" % ("OFF (--insecure)" if opts.insecure else "on"))
    print("In the generator set Base URL to:  http://%s:%d/v1"
          % ("localhost" if opts.host == "0.0.0.0" else opts.host, opts.port))
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        httpd.shutdown()


if __name__ == "__main__":
    main()
